### ISSUE TITLE
Unable to build rules_haskell in a docker container using nix >= 2.2

### ISSUE THREAD

**Issue by NinjaTrappeur** [tweag/rules_haskell#591]
**TL;DR**: There's a weird sandboxing error when building rules_haskell in a docker -> Bazel Sandbox -> Nix Sandbox stack. This error is not reproducible using rules_nixpkgs alone, it only appears when adding the `all-haskell-packages.bzl.drv` derivation to the store. I cannot find a minimal repro, I need help on this.

====

Continuation of https://github.com/tweag/rules_haskell/pull/574#issuecomment-455171557 .

The nixpkgs repository recently upgraded nix to 2.2. While this transition does not affect the rules_haskell build on a standard environment, it does breaks the circleci and any docker-based build.

The build usually fails with a:

```
error: while setting up the build environment: unable to make '/' private mount: Permission denied
```
or
```
error: cloning builder process: Operation not permitted
```

Depending on the host docker version.

Nix 2.2 enabled sandboxing by default on linux: this seems to be the problem's root cause. However, I'm not sure whether this is:

1. A nix bug.
1. A rules_haskell bug.
1. A docker bug.

In order to find the real culprit, we should find a way to minimize the error case. However, I have not been able to reproduce this outside of this very same configuration.

I first though the problem was coming from rules_nixpkgs and tried to trigger the same error without rules_haskell without success. https://github.com/NinjaTrappeur/nix22sandboxerror is not triggering the error and succeed.

Overall, it seems this problem is not triggered as long as we don't use the `haskell_nixpkgs_packages` rule with nix >= 2.2.

# Steps to reproduce

In order to reproduce the issue, you'll need docker, a nixpkgs local checkout and a rules_haskell_checkout.

- docker
- A nixpkgs clone set to the revision `da8e0491bc9bd9ee193e71776dc7279b2e5efef3`
- A rules_haskell clone set to the revision `d2abf00e76fe055c19cdb35dc30c16478e06d072`

Apply

```diff
diff --git a/nixpkgs/default.nix b/nixpkgs/default.nix
index 639b1cd..dffef0c 100644
--- a/nixpkgs/default.nix
+++ b/nixpkgs/default.nix
@@ -1,5 +1 @@
-import (fetchTarball {
-   # Nixpkgs checkout from 2018-11-03.
-   url = "https://github.com/NixOS/nixpkgs/archive/cb692e6f52f6d4ca619c9fbc7a80afc081a93103.tar.gz";
-   sha256 = "1aww7wjgv868asd4yl0wnzh8qa26j4wsmqn85b3s9fkqkz4p2m0f";
-})
+import /nixpkgs/default.nix
```

To make your rules_haskell checkout use our local nixpkgs checkout.

Finally, run

```
docker run -d -it --name circledebug --mount type=bind,source=${nixpkgs_checkout}=/nixpkgs --mount type=bind,source=${rules_haskell_checkout},target=/rules_haskell nixos/nix:2.1.3
```

# Potential Fix

Disabling the nix-build sandbox in rules_nixpkgs fix the problem.

```diff
diff --git a/nixpkgs/nixpkgs.bzl b/nixpkgs/nixpkgs.bzl
index 72df34d..046d811 100644
--- a/nixpkgs/nixpkgs.bzl
+++ b/nixpkgs/nixpkgs.bzl
@@ -104,6 +104,7 @@ def _nixpkgs_package_impl(repository_ctx):
         # the store path.
         "--out-link",
         "bazel-support/nix-out-link",
+        "--option", "sandbox", "false"
     ])
 
     expr_args.extend(repository_ctx.attr.nixopts)
```

We may additionally add a `--define disable_nix_sandbox=true` flag to let nix use a sandbox in most cases while disabling it on circle-ci.

However, I'm not sure this is an acceptable solution.

**Comment by NinjaTrappeur** [comment]
Apparently, only the `haskell_nixpkgs_package`, `nixpkgs_cc_configure`, `haskell_nixpkgs_packageset` and the `@nixpkgs_nodejs` were triggering this issue. The other packages can be built from within the nix sandbox.

Strange.

**Comment by mboes** [comment]
What's left to be done on this ticket?

**Comment by NinjaTrappeur** [comment]
1. We need to be able to evaluate the `WORKSPACE` when nix is not in path. https://github.com/tweag/rules_nixpkgs/pull/64 (edit: just has been merged while I was typing this)
2. Pin a new rules_nixpkgs release and send a PR for https://github.com/tweag/rules_haskell/tree/nin-refactor-unsupported-platforms-check

**Comment by Profpatsch** [comment]
@NinjaTrappeur The [commit of 2](https://github.com/tweag/rules_nixpkgs/pull/64/commits/27df5568b8b819145dd54bc389bb4f57a18b8e00) is in `rules_nixpkgs` 0.5.2 now, so this issue is unblocked.

**Comment by mboes** [comment]
@NinjaTrappeur I believe this can be closed now. Reopen if not.
